fix FreeRDP some memory leak problem

weizhenwei opened the pull request:
using valgrind to check these memory leak, and fix them.

bmiklautz commented:
@freerdp-bot test

bmiklautz commented:
@weizhenwei thanks. Would you update your pull request to not contain merges?

weizhenwei commented:
remove all merge commits, and fix compile error on win64/vs2010 platform.

weizhenwei commented:
@bmiklautz , please help to check.

bmiklautz commented:
@freerdp-bot test

weizhenwei commented:
I checked Jenkins output: "[WS-CLEANUP] Deleting project workspace...
ERROR: [WS-CLEANUP] Cannot delete workspace: remote file operation failed: c:\Devel\workspace\windows_msbuild\arch\win32\label\vs2013 at hudson.remoting.Channel@4429bfe2:JNLP4-connect connection from frdp-win7.in.freerdp.com/10.20.50.14:49168: java.io.IOException: Unable to delete 'c:\Devel\workspace\windows_msbuild\arch\win32\label\vs2013\build\CMakeCache.txt'. Tried 3 times (of a maximum of 3) waiting 0,1 sec between attempts."  What could I do about this?

bmiklautz commented:
@weizhenwei I'll have a look.

hardening commented on libfreerdp/core/update.c:
```diff
@@ -2139,6 +2139,17 @@ void update_free(rdpUpdate* update)
 		free(update->secondary);
 		free(update->altsec);
+		free(update->window->monitored_desktop.windowIds);
+		free(update->window->window_state.titleInfo.string);
```
may I suggest to create function to free the window_state (it could be used below) ?

weizhenwei replied:
OK, I will create a function to free window_state and call it at multiple place.

hardening commented on libfreerdp/core/window.c:
```diff
@@ -350,6 +352,14 @@ BOOL update_read_window_state_order(wStream* s, WINDOW_ORDER_INFO* orderInfo, WI
 {
+	if (window_icon->iconInfo)
+	{
+		free(window_icon->iconInfo->bitsColor);
```
same remark here, it would be neat to have a function to free the icon info and also use it above.

bmiklautz commented:
@freerdp-bot test.

weizhenwei commented:
@hardening , @bmiklautz , I refined the patch to remove redundant code copy, please check, thank you very much.

ilammy commented on winpr/libwinpr/file/generic.c:
```diff
@@ -938,6 +938,7 @@ HANDLE FindFirstFileW(LPCWSTR lpFileName, LPWIN32_FIND_DATAW lpFindFileData)
 
+	free(fd);
 	return h;
 }
```
It's a matter of style, but with functions that deal with resources I personally like to have a single return point and structure error handling with goto. This ensures that the resources are always freed.

Like this:
```diff
 	if (h != INVALID_HANDLE_VALUE)
 	{
 		if (!ConvertFindDataAToW(fd, lpFindFileData))
 		{
 			SetLastError(ERROR_NOT_ENOUGH_MEMORY);
-			free(fd);
-			return INVALID_HANDLE_VALUE;
+			h = INVALID_HANDLE_VALUE;
+			goto out;
  		}
  	}
-
+out:
+	free(fd);
  	return h;
  }
```
Though maybe it's an overkill in this case.

ilammy commented on channels/rail/client/rail_orders.c:
```diff
@@ -650,7 +650,9 @@ UINT rail_order_recv(railPlugin* rail, wStream* s)
 				RAIL_EXEC_RESULT_ORDER execResult;
-				return rail_recv_exec_result_order(rail, &execResult, s);
+				error = rail_recv_exec_result_order(rail, &execResult, s);
+				free(execResult.exeOrFile.string);
```
This field is not actually initialized in execResult. If `rail_recv_exec_result_order()` fails before filling in the string (e.g., in `rail_read_server_exec_result_order()`) the we'll get heap corruption here.

I'd recommend to zero-initialize execResult (and other structures in this function, while we're at it). I guess you can try adding `= { 0 }` to declarations, but if that fails then explicit `ZeroMemory` calls should do the trick.

ilammy replied:
Oh, nevermind. I missed the `ZeroMemory` call in rail_recv_exec_result_order(). It is too sneaky placed near the variable declarations.

So there are no errors here, the free() call is fine.

ilammy commented on libfreerdp/core/window.c:
```diff
@@ -294,6 +294,7 @@ BOOL update_read_window_state_order(wStream* s, WINDOW_ORDER_INFO* orderInfo, WI
 
+		free(windowState->windowRects);
 		size = sizeof(RECTANGLE_16) * windowState->numWindowRects;
 		windowState->windowRects = (RECTANGLE_16*) malloc(size);
```
I guess we can do realloc() here instead of free() + malloc(). You have to be careful to handle realloc() failure, but I think this field will be eventually freed, so we can simply return here if realloc fails, just as we do now for malloc().

weizhenwei replied:
I took reference on other realloc() call, and rewrote like this:
		size = sizeof(RECTANGLE_16) * windowState->numWindowRects;
		newRect = (RECTANGLE_16*)realloc(windowState->windowRects, size);
		if (!newRect)
		{
			free(windowState->windowRects);
			windowState->windowRects = NULL;
			return FALSE;
		}
		windowState->windowRects = newRect;

ilammy commented on libfreerdp/core/window.c:
```diff
@@ -328,6 +329,7 @@ BOOL update_read_window_state_order(wStream* s, WINDOW_ORDER_INFO* orderInfo, WI
 
+		free(windowState->visibilityRects);
 		size = sizeof(RECTANGLE_16) * windowState->numVisibilityRects;
 		windowState->visibilityRects = (RECTANGLE_16*) malloc(size);
```
Here is the same realloc() thing as above.

weizhenwei commented on channels/rail/client/rail_main.c:
```diff
@@ -620,9 +620,11 @@ static void* rail_virtual_channel_client_thread(void* arg)
 				WLog_ERR(TAG, "rail_order_recv failed with error %"PRIu32"!", error);
 				break;
 			}
+			Stream_Free(data, TRUE);
```
Thank you for  the suggestion, it's good to me, I will refactor to avoid dulicating Stream_Free.

akallabeth commented on channels/drive/client/drive_main.c:
```diff
@@ -867,6 +868,7 @@ UINT drive_register_drive_path(PDEVICE_SERVICE_ENTRY_POINTS pEntryPoints,
 out_error:
 	MessageQueue_Free(drive->IrpQueue);
 	ListDictionary_Free(drive->files);
+	free(drive->path);
```
This looks broken in context. Should use ```drive_free``` to remove duplication (and possible missing cleanup)

weizhenwei replied:
I will replace free(drive->path) with drive_free(drive)  and test.

akallabeth commented on channels/rail/client/rail_main.c:
```diff
@@ -92,6 +92,17 @@ UINT rail_send_channel_data(railPlugin* rail, void* data, size_t length)
+ */
+static void rail_client_clean_exec_order(RAIL_EXEC_ORDER* exec)
+{
+	free(exec->exeOrFile.string);
```
Missing NULL pointer check, set pointers to ```NULL``` afterward.

akallabeth commented on libfreerdp/core/update.c:
```diff
@@ -2041,6 +2041,13 @@ static void update_free_queued_message(void* obj)
 
+static void update_free_window_state(WINDOW_STATE_ORDER window_state)
+{
+	free(window_state.titleInfo.string);
```
Would suggest makeing ```window_state``` a pointer and set it to ```NULL``` after free. That way an invalid access will crash the application reliably.

akallabeth commented on libfreerdp/core/window.c:
```diff
@@ -581,6 +597,14 @@ BOOL update_recv_desktop_info_order(rdpUpdate* update, wStream* s, WINDOW_ORDER_
 
+void update_free_window_icon_info(ICON_INFO* iconInfo)
+{
+	free(iconInfo->bitsColor);
```
Missing NULL pointer check. Set freed pointers to NULL afterward

weizhenwei replied:
I will do all theNULL pointer check and set freed pointers to NULL afterward.

mfleisz commented on channels/rail/client/rail_main.c:
```diff
@@ -97,9 +97,26 @@ UINT rail_send_channel_data(railPlugin* rail, void* data, size_t length)
+    if (!exec)
+        return;
+
+    if (exec->exeOrFile.string)
```
NULL checks before free() are useless because free() already does that check

weizhenwei replied:
as suggested, I added these NULL checks. please give some tips if there are better ways to do so.

akallabeth replied:
As @mfleisz mentioned, not necessary to check the pointers to be freed, only the data structure dereferenced.

weizhenwei commented:
@akallabeth , I added all the  NULL pointer check and set freed pointers to NULL afterward, please help to review.

akallabeth commented on channels/drive/client/drive_main.c:
```diff
@@ -868,7 +868,7 @@ UINT drive_register_drive_path(PDEVICE_SERVICE_ENTRY_POINTS pEntryPoints,
 	ListDictionary_Free(drive->files);
-	free(drive->path);
+	drive_free(drive);
 	free(drive);
```
double free now (drive is already freed in ```drive_free```)

weizhenwei replied:
sorry for my stupid fault.

akallabeth commented on libfreerdp/core/update.c:
```diff
@@ -2041,11 +2041,28 @@ static void update_free_queued_message(void* obj)
+    if (!window_state)
+        return;
+
+    if (window_state->titleInfo.string)
```
Same as above.

akallabeth commented on libfreerdp/core/window.c:
```diff
@@ -599,9 +600,24 @@ BOOL update_recv_desktop_info_order(rdpUpdate* update, wStream* s, WINDOW_ORDER_
-	free(iconInfo->bitsColor);
-	free(iconInfo->bitsMask);
-	free(iconInfo->colorTable);
+    if (iconInfo->bitsColor)
```
same as above

weizhenwei commented:
@akallabeth , @mfleisz , I removed the uselsess NULL pointer check and rewrote the drive_free() call, please help to check. Thank you for your help and patience.

akallabeth commented:
@freerdp-bot test

bmiklautz commented on libfreerdp/core/window.c:
```diff
@@ -329,9 +335,14 @@ BOOL update_read_window_state_order(wStream* s, WINDOW_ORDER_INFO* orderInfo, WI
-		windowState->visibilityRects = (RECTANGLE_16*) malloc(size);
-		if (!windowState->visibilityRects)
+		newRect = (RECTANGLE_16*)realloc(windowState->visibilityRects, size);
+		if (!newRect)
```
This can cause a double free if windowState->visibilityRects is 0.

bmiklautz replied:
@weizhenwei no problem ;) - We are happy about the improvements. Just wanted to mention it that we have it in one place.

bmiklautz commented on libfreerdp/core/window.c:
```diff
@@ -295,9 +296,14 @@ BOOL update_read_window_state_order(wStream* s, WINDOW_ORDER_INFO* orderInfo, WI
-		windowState->windowRects = (RECTANGLE_16*) malloc(size);
-		if (!windowState->windowRects)
+		newRect = (RECTANGLE_16*)realloc(windowState->windowRects, size);
+		if (!newRect)
```
This can cause a double free if windowState->numWindowRects is 0. (And also breaks rails).
